File: src/bo4e_cli/io/git.py
"""
This module provides functions to interact with the git repository in the current working directory.
It is designed to interact with the BO4E repository in order to retrieve version tags, branches, and commits.
"""

import re
import subprocess
from typing import Iterable, Literal

from more_itertools import one

from bo4e_cli.io.console import CONSOLE
from bo4e_cli.models.version import Version


def is_version_tag(value: str) -> bool:
    """
    Check if value is a valid version tag and exists in repository.
    """
    try:
        Version.from_str(value)
        subprocess.check_call(["git", "show-ref", "--quiet", f"refs/tags/{value}"])
    except (ValueError, subprocess.CalledProcessError):
        return False
    return True


def is_branch(value: str) -> bool:
    """
    Check if a branch is a valid branch name and exists in repository.
    """
    try:
        subprocess.check_call(["git", "show-ref", "--quiet", f"refs/remotes/origin/{value}"])
        return True
    except subprocess.CalledProcessError:
        return False


def get_branches_containing_commit(commit_id: str) -> Iterable[str]:
    """
    Get all branches containing the commit id.
    If the commit id is not found, a subprocess.CalledProcessError will be raised.
    If the commit exists but is not on any branch (e.g. only on tags), an empty Iterable will be returned.
    """
    cmd = ["git", "branch", "-a", "--contains", commit_id]
    output = subprocess.check_output(cmd).decode().strip()
    if output.startswith("error: no such commit"):
        raise subprocess.CalledProcessError(1, cmd, output=output)
    return (line.strip().lstrip("*").lstrip() for line in output.splitlines())


def is_commit(value: str) -> bool:
    """
    Check if value is a valid commit id.
    """
    try:
        if re.fullmatch(r"^[0-9a-f]{40}$", value) is None:
            return False
        _ = get_branches_containing_commit(value)
        # If the commit ID doesn't exist, an error will be raised.
    except subprocess.CalledProcessError:
        return False
    return True


def get_checkout_commit_id() -> str:
    """
    Get the commit id of the current checkout.
    """
    return subprocess.check_output(["git", "rev-parse", "HEAD"]).decode().strip()


def _get_ref(ref: str) -> tuple[Literal["tag", "branch", "commit"], str]:
    """
    Get the type of reference and the reference itself.
    """
    if is_version_tag(ref):
        CONSOLE.print(f"Get tags before tag {ref}", show_only_on_verbose=True)
        return "tag", ref
    if is_branch(ref):
        CONSOLE.print(f"Get tags on branch {ref}", show_only_on_verbose=True)
        return "branch", ref
    if is_commit(ref):
        CONSOLE.print(f"Get tags before commit {ref}", show_only_on_verbose=True)
        return "commit", ref
    cur_commit = get_checkout_commit_id()
    CONSOLE.print(
        f"Supplied value ({ref}) is neither a tag, a branch nor a commit. "
        f"Get tags before current checkout commit {cur_commit}",
        show_only_on_verbose=True,
    )
    return "commit", cur_commit


def get_last_n_tags(
    n: int, *, ref: str = "main", exclude_candidates: bool = True, exclude_technical_bumps: bool = False
) -> Iterable[str]:
    """
    Get the last n tags in chronological descending order starting from `ref`.
    If `ref` is a branch, it will start from the current HEAD of the branch.
    If `ref` is a tag, it will start from the tag itself. But the tag itself will not be included in the output.
    If `ref` is neither nor, the main branch will be used as fallback.
    If `exclude_candidates` is True, candidate versions will be excluded from the output.
    If the number of found versions is less than `n`, a warning will be logged.
    If n=0, all versions since v202401.0.0 will be taken into account.
    If exclude_technical_bumps is True, from each functional release group,
    the highest technical release will be returned.
    """
    version_threshold = "v202401.0.0"  # Is used if n=0
    ref_type, reference = _get_ref(ref)
    if n == 0:
        CONSOLE.print(f"Get all tags since {version_threshold}", show_only_on_verbose=True)
    else:
        CONSOLE.print(f"Get the last {n} tags", show_only_on_verbose=True)

    CONSOLE.print(f"{'Exclude' if exclude_candidates else 'Include'} release candidates", show_only_on_verbose=True)
    CONSOLE.print(f"{'Exclude' if exclude_technical_bumps else 'Include'} technical bumps", show_only_on_verbose=True)
    output = (
        subprocess.check_output(["git", "tag", "--merged", reference, "--sort=-creatordate"])
        .decode()
        .strip()
        .splitlines()
    )
    if len(output) == 0:
        CONSOLE.print("No tags found.", style="warning")
        return
    last_version = Version.from_str(output[0])

    counter = 0
    stop_iteration = False
    for ind, tag in enumerate(output):
        if counter >= n > 0:
            stop_iteration = True
        if stop_iteration:
            return
        if n == 0 and tag == version_threshold:
            stop_iteration = True
        version = Version.from_str(tag)
        # pylint: disable=too-many-boolean-expressions
        if (
            exclude_candidates
            and version.is_release_candidate()
            or exclude_technical_bumps
            and ind > 0
            and last_version.bumped_technical(version)
            or ind == 0
            and ref_type == "tag"
        ):
            CONSOLE.print(f"Skipping version {version}", show_only_on_verbose=True)
            continue
        CONSOLE.print(f"Yielding version {version}", show_only_on_verbose=True)
        yield tag
        last_version = version
        counter += 1
    if counter < n and 0 < n:
        if ref_type == "tag":
            CONSOLE.print(f"Only found {counter} tags before tag {ref}, tried to retrieve {n}", style="warning")
        else:
            CONSOLE.print(f"Only found {counter} tags on branch {ref}, tried to retrieve {n}", style="warning")
    if n == 0:
        CONSOLE.print(f"Threshold version {version_threshold} not found. Returned all tags.", style="warning")


def get_last_version_before(version: Version) -> Version:
    """
    Get the last non-candidate version before the provided version following the commit history.
    """
    return Version.from_str(one(get_last_n_tags(1, ref=str(version))))

File: src/io/git.rs
use crate::models::git::{RefKind, Reference};
use crate::models::version::Version;
use crate::repo::filter::{FilterOptions, filter_tags};
use std::io;
use std::path::Path;
use std::process::{Command, Output};
use std::str::FromStr;

fn check_success(output: &Output, error_message: &str) -> io::Result<()> {
    if !output.status.success() {
        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);
        Err(io::Error::new(
            io::ErrorKind::Other,
            format!("{error_message}\nStdout: {stdout}\nStderr: {stderr}"),
        ))
    } else {
        Ok(())
    }
}

#[allow(dead_code)]
pub fn clone_repo(repo_url: &str, branch_or_tag: &str, dest: &Path) -> io::Result<()> {
    let output = Command::new("git")
        .args(["clone", "--branch", branch_or_tag, "--depth", "1", repo_url])
        .arg(dest.as_os_str())
        .output()?; // get exit status

    check_success(&output, "Failed to clone repository.")
}

pub fn is_version_tag(value: &str) -> io::Result<bool> {
    Command::new("git")
        .args(["show-ref", "--quiet", &format!("refs/tags/{value}")])
        .status()
        .map(|exit_status| exit_status.success())
}

pub fn is_branch(value: &str) -> io::Result<bool> {
    // Check local branches first, then remote-tracking branches.
    let local = Command::new("git")
        .args(["show-ref", "--quiet", &format!("refs/heads/{value}")])
        .status()
        .map(|s| s.success())?;
    if local {
        return Ok(true);
    }
    Command::new("git")
        .args([
            "show-ref",
            "--quiet",
            &format!("refs/remotes/origin/{value}"),
        ])
        .status()
        .map(|exit_status| exit_status.success())
}

/// Returns true if `value` resolves to a commit (any ref `git branch --contains` accepts).
///
/// Note: this returns true for tag and branch names too. Call this after
/// `is_version_tag` and `is_branch` if you want strict "is this a raw commit hash" classification.
pub fn is_commit_hash(value: &str) -> io::Result<bool> {
    match get_branches_containing_commit(value) {
        Ok(_) => Ok(true),
        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(false),
        Err(e) => Err(e),
    }
}

fn get_branches_containing_commit(commit: &str) -> io::Result<Vec<String>> {
    let output = Command::new("git")
        .args(["branch", "-a", "--contains", commit])
        .output()?;

    // `git branch --contains` may print "error: …" to stderr and exit non-zero for
    // unknown/malformed refs — treat these as "not a commit" rather than a hard error.
    if !output.status.success() {
        return Err(io::Error::new(
            io::ErrorKind::NotFound,
            "No branches found containing the specified commit.",
        ));
    }
    let output = String::from_utf8_lossy(&output.stdout);
    let output = output.trim();
    if output.starts_with("error: no such commit") {
        return Err(io::Error::new(
            io::ErrorKind::NotFound,
            "No branches found containing the specified commit.",
        ));
    }
    Ok(output
        .lines()
        .map(|line| line.trim().trim_start_matches('*').trim_start().to_string())
        .collect())
}

pub fn get_commit_sha(branch_or_tag: &str) -> io::Result<String> {
    let output = Command::new("git")
        .args(["rev-parse", branch_or_tag])
        .output()?;

    check_success(&output, "Failed to get commit SHA.")?;
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

pub fn get_commit_date(commit: &str) -> io::Result<String> {
    let output = Command::new("git")
        .args(["show", "-s", "--format=%ci", commit])
        .output()?;

    check_success(&output, "Failed to get commit date.")?;
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

#[allow(dead_code)]
fn parse_reference(value: String) -> io::Result<Reference> {
    if is_version_tag(&value)? {
        Ok(Reference::Tag(value))
    } else if is_branch(&value)? {
        Ok(Reference::Branch(value))
    } else if is_commit_hash(&value)? {
        Ok(Reference::Commit(value))
    } else if value == "HEAD" {
        Ok(Reference::Head)
    } else {
        Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Supplied value is not a valid branch, tag, commit hash or the string literal 'HEAD'.",
        ))
    }
}

pub struct GetLastNTagsOpts<'a, F>
where
    F: FnMut(&Version) -> Result<bool, String>,
{
    pub n: u32,
    pub reference: &'a str,
    pub exclude_candidates: bool,
    pub exclude_technical_bumps: bool,
    pub skip_first: bool,
    pub is_release: F,
}

pub fn get_last_n_tags<F>(opts: GetLastNTagsOpts<'_, F>) -> Result<Vec<Version>, String>
where
    F: FnMut(&Version) -> Result<bool, String>,
{
    let raw = tags_merged(opts.reference).map_err(|e| e.to_string())?;

    let mut candidates: Vec<Version> = Vec::with_capacity(raw.len());
    for tag in raw {
        match Version::from_str(&tag) {
            Ok(v) => candidates.push(v),
            Err(_) => crate::cwarn!("skipping unparseable tag '{tag}'"),
        }
    }

    let filter_opts = FilterOptions {
        n: opts.n,
        exclude_candidates: opts.exclude_candidates,
        exclude_technical_bumps: opts.exclude_technical_bumps,
        skip_first: opts.skip_first,
        threshold: Version::from_str("v202401.0.0").expect("hardcoded threshold parses"),
    };

    filter_tags(&candidates, &filter_opts, opts.is_release)
}

pub fn tags_merged(reference: &str) -> io::Result<Vec<String>> {
    let output = Command::new("git")
        .args([
            "tag",
            "--merged",
            reference,
            "--sort=-version:refname",
            "--sort=-creatordate",
        ])
        .output()?;
    check_success(&output, "Failed to list merged tags.")?;
    Ok(String::from_utf8_lossy(&output.stdout)
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
        .map(String::from)
        .collect())
}

/// Classify a user-supplied ref string as a tag, branch, or commit, and resolve it.
///
/// Precedence (when a value matches multiple): tag → branch → commit. If the value
/// resolves to none of these, falls back to the current HEAD's SHA and emits an
/// info message via `cprint_normal!`.
///
/// For `RefKind::Commit`, the returned string is always a concrete 40-char SHA —
/// shorthand like `HEAD` or `HEAD~3` is resolved via `git rev-parse`.
pub fn get_ref(value: &str) -> io::Result<(RefKind, String)> {
    if is_version_tag(value)? {
        return Ok((RefKind::Tag, value.to_string()));
    }
    if is_branch(value)? {
        return Ok((RefKind::Branch, value.to_string()));
    }
    if is_commit_hash(value)? {
        return Ok((RefKind::Commit, get_commit_sha(value)?));
    }
    let cur = get_commit_sha("HEAD")?;
    let short: String = cur.chars().take(7).collect();
    crate::cprint_normal!("'{value}' is not a tag, branch, or commit; falling back to HEAD ({short}).");
    Ok((RefKind::Commit, cur))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::process::Command;
    use std::sync::Mutex;

    // Serializes tests that mutate process cwd. Cargo runs tests in parallel by default;
    // any test in this module that calls set_current_dir must hold this lock.
    static CWD_LOCK: Mutex<()> = Mutex::new(());

    /// Initialize a git repo with 3 tagged commits, acquire the cwd lock, and set
    /// the process cwd to the tempdir. Returns both so the caller holds them for the
    /// test's full lifetime — dropping either ends the exclusive window.
    fn make_git_repo() -> (tempfile::TempDir, std::sync::MutexGuard<'static, ()>) {
        let guard = CWD_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        let dir = tempfile::tempdir().unwrap();
        let p = dir.path();
        let run = |args: &[&str]| {
            let out = Command::new("git")
                .args(args)
                .current_dir(p)
                .output()
                .expect("git invocation failed");
            assert!(
                out.status.success(),
                "git {args:?} failed: {}",
                String::from_utf8_lossy(&out.stderr)
            );
        };
        run(&["init", "-q", "-b", "main"]);
        run(&["config", "user.email", "t@t.t"]);
        run(&["config", "user.name", "t"]);
        run(&["commit", "--allow-empty", "-m", "c1", "-q"]);
        run(&["tag", "v202401.0.1"]);
        run(&["commit", "--allow-empty", "-m", "c2", "-q"]);
        run(&["tag", "v202401.0.2"]);
        run(&["commit", "--allow-empty", "-m", "c3", "-q"]);
        run(&["tag", "v202401.1.0"]);
        run(&["tag", "not-a-version"]);
        std::env::set_current_dir(p).unwrap();
        (dir, guard)
    }

    #[test]
    fn test_tags_merged_returns_descending_version_order() {
        let (_dir, _guard) = make_git_repo();
        let tags = tags_merged("HEAD").unwrap();
        // Descending: 1.0 first, then 0.2, 0.1, then non-version tag.
        assert_eq!(tags[0], "v202401.1.0");
        assert!(tags.contains(&"v202401.0.2".to_string()));
        assert!(tags.contains(&"v202401.0.1".to_string()));
        assert!(tags.contains(&"not-a-version".to_string()));
    }

    #[test]
    fn test_get_ref_classifies_tag_branch_and_falls_back_to_head() {
        let (_dir, _guard) = make_git_repo();

        // Existing console for the fallback's info message.
        use crate::console::console::{Console, Level, CONSOLE};
        let _ = CONSOLE.set(Console::new(Level::Quiet));

        let (kind, value) = get_ref("v202401.0.1").unwrap();
        assert_eq!(kind, crate::models::git::RefKind::Tag);
        assert_eq!(value, "v202401.0.1");

        let (kind, value) = get_ref("main").unwrap();
        assert_eq!(kind, crate::models::git::RefKind::Branch);
        assert_eq!(value, "main");

        // Unknown value → fallback to HEAD's commit SHA.
        let (kind, value) = get_ref("definitely-not-a-ref").unwrap();
        assert_eq!(kind, crate::models::git::RefKind::Commit);
        assert_eq!(value.len(), 40); // full SHA from `git rev-parse HEAD`

        // "HEAD" should resolve to a concrete SHA, not pass through as a literal.
        let (kind, value) = get_ref("HEAD").unwrap();
        assert_eq!(kind, crate::models::git::RefKind::Commit);
        assert_eq!(value.len(), 40);
        assert!(value.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn test_get_last_n_tags_returns_valid_versions_in_order() {
        let (_dir, _guard) = make_git_repo();

        use crate::console::console::{Console, Level, CONSOLE};
        let _ = CONSOLE.set(Console::new(Level::Quiet));

        let opts = GetLastNTagsOpts {
            n: 0,
            reference: "HEAD",
            exclude_candidates: false,
            exclude_technical_bumps: false,
            skip_first: false,
            is_release: |_| Ok(true),
        };
        let out = get_last_n_tags(opts).unwrap();
        // The 'not-a-version' tag is dropped; 3 valid versions remain in descending order.
        assert_eq!(out.len(), 3);
        assert_eq!(out[0].to_string(), "v202401.1.0");
        assert_eq!(out[1].to_string(), "v202401.0.2");
        assert_eq!(out[2].to_string(), "v202401.0.1");
    }
}
